### PR TITLE
Add armhf support for the forthcoming K8s release

### PR DESCRIPTION
**Pull request by mattmoor**
cc @hasheddan

This still needs the cloudbuild changes, which I'll stage and test ~now.

### CODE REVIEW

**Comment by mattmoor** [comment]
Alright, I just did a GCB run with the changes I'm about to push, and this is one of the resulting images:
```
crane manifest gcr.io/mattmoor-knative/static:latest | jq .
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
  "manifests": [
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 426,
      "digest": "sha256:a7fbf4bc518870042021d9fff54b486da3c3182f880b52f91439d63895ec8423",
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 426,
      "digest": "sha256:0fb03485bed51c9ee7a7c7362da40c23e202ec75f8660a5fa3b6164849866dc5",
      "platform": {
        "architecture": "arm",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 426,
      "digest": "sha256:fcee51c84824bea83e2692cd1fb2a5d26038371581b2bef80a167a374a01f35e",
      "platform": {
        "architecture": "arm64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 426,
      "digest": "sha256:6284b3a8dbdb0df27072217be18115224c392df29d2cb79c5c3cbef72a33e649",
      "platform": {
        "architecture": "ppc64le",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 426,
      "digest": "sha256:b0dc5adcc32d2b00f871ea0d4badd0d7bd884255a1b3f47b199fd40ad03a7f07",
      "platform": {
        "architecture": "s390x",
        "os": "linux"
      }
    }
  ]
}
```

**Comment by chanseokoh** [comment]
There are build failures. Seems related?

**Comment by mattmoor** [comment]
@chanseokoh I must've messed something up in my second commit Friday, I'll TAL now.

**Comment by mattmoor** [comment]
Looks like just a broken label in `examples/python3/BUILD`, should be fixed now 🤞

File: base/distro.bzl [hunk 1]
```diff
@@ -1,5 +1,7 @@
 load("@package_bundle_amd64_debian9//file:packages.bzl", packages_amd64_debian9 = "packages")
 load("@package_bundle_amd64_debian10//file:packages.bzl", packages_amd64_debian10 = "packages")
+load("@package_bundle_arm64_debian9//file:packages.bzl", packages_arm_debian9 = "packages")
+load("@package_bundle_arm64_debian10//file:packages.bzl", packages_arm_debian10 = "packages")
```
Comment (chanseokoh) [inline base/distro.bzl:4]: Should it be like this?
```suggestion
load("@package_bundle_arm_debian9//file:packages.bzl", packages_arm_debian9 = "packages")
load("@package_bundle_arm_debian10//file:packages.bzl", packages_arm_debian10 = "packages")
```

**Comment by chanseokoh** [comment]
Oh, and please resolve the merge conflict.

File: base/distro.bzl (hunk 1 above)
Comment (mattmoor) [reply base/distro.bzl:4]: good catch 😅

**Comment by mattmoor** [comment]
I'm also going to see what's left to move this to `BASE_ARCHITECTURES` now that I have the rest working, but running updateWorkspaceSnapshots.sh takes a really long time 😅

**Comment by mattmoor** [comment]
Pushing the changed now.  I'm going to submit a fresh GCB run to confirm things work there now.

**Comment by mattmoor** [comment]
I forgot about the `ld.so.{arch}.cache` stuff, so until we have arm hardware to generate that on, we can't easily add Python support.  This basically extends to all architectures that we don't have *CI* for because the verification that the cache is correct won't be running and we'll end up shipping stale files 😬 

I've dropped `arm` from BASE for now, to try and unblock Kubernetes, and have another GCB run going with that to verify the latest version of these changes.

**Comment by mattmoor** [comment]
@chanseokoh this should be RFAL.

**Comment by chanseokoh** [comment]
Thanks for thoroughly testing this! I'll try to get back to this by EOD.

File: BUILD [hunk 2]
```diff
@@ -1,241 +1,280 @@
...
+        ("latest", "python3", "root"),
+        ("nonroot", "python3", "nonroot"),
+        ("debug", "debug", "root"),
+        ("debug-nonroot", "debug", "nonroot"),
+    ]
+})
+
+PYTHON3.update({
+    "gcr.io/{PROJECT_ID}/python3-" + distro + ":" + tag_base + "-" + arch: "//experimental/python3:" + label + "_" + user + "_" + arch + "_" + distro
+    for arch in BASE_ARCHITECTURES
```
Comment (chanseokoh) [inline BUILD:112]: Image tags in `python3-debian{9|10}` (as opposed to `python3`) were all amd64 images and not manifest lists. This code now pushes multiple architectures, which is fine. But probably you'll want to add `docker create manifest` in `cloudbuild_docker.sh`. Also worth updating the PR title to also add "convert `python{3|2.7}-debian{9|10}` multi-arch" for easy reference in the future.

File: BUILD [hunk 3]
```diff
@@ -1,241 +1,280 @@
...
+    "gcr.io/{PROJECT_ID}/python2.7:" + tag_base + "-" + arch: "//experimental/python2.7:" + label + "_" + arch + "_debian9"
+    for arch in BASE_ARCHITECTURES
+    for (tag_base, label) in [
+        ("latest", "python27"),
+        ("debug", "debug"),
+    ]
+})
+
+PYTHON27.update({
+    "gcr.io/{PROJECT_ID}/python2.7-" + distro + ":" + tag_base + "-" + arch: "//experimental/python2.7:" + label + "_" + arch + "_" + distro
```
Comment (chanseokoh) [inline BUILD:138]: ditto for `python2.7-debian{9|10}`.

File: BUILD [hunk 4]
```diff
@@ -1,241 +1,280 @@
...
+        ("debug", "debug", "root"),
+        ("debug-nonroot", "debug", "nonroot"),
+    ]
+    for distro in DISTROS
+})
+
+PYTHON3 = {
+    "gcr.io/{PROJECT_ID}/python3:{COMMIT_SHA}": "//experimental/python3:python3_root_amd64_debian9",
+    "gcr.io/{PROJECT_ID}/python3-debian9:{COMMIT_SHA}": "//experimental/python3:python3_root_amd64_debian9",
+    "gcr.io/{PROJECT_ID}/python3-debian10:{COMMIT_SHA}": "//experimental/python3:python3_root_amd64_debian10",
```
Comment (chanseokoh) [inline BUILD:96]: Let's not add these. The current Python images don't have these tags. I actually plan to remove `{COMMIT_SHA}` from other images as well.

File: BUILD [hunk 5]
```diff
@@ -1,241 +1,280 @@
...
+        ("debug", "debug", "root"),
+        ("debug-nonroot", "debug", "nonroot"),
+    ]
+    for distro in DISTROS
+})
+
+PYTHON27 = {
+    "gcr.io/{PROJECT_ID}/python2.7:{COMMIT_SHA}": "//experimental/python2.7:python27_amd64_debian9",
+    "gcr.io/{PROJECT_ID}/python2.7-debian9:{COMMIT_SHA}": "//experimental/python2.7:python27_amd64_debian9",
+    "gcr.io/{PROJECT_ID}/python2.7-debian10:{COMMIT_SHA}": "//experimental/python2.7:python27_amd64_debian10",
```
Comment (chanseokoh) [inline BUILD:125]: ditto

File: BUILD (hunk 4 above)
Comment (mattmoor) [reply BUILD]: Remove the COMMIT_SHA images from both pythons

File: BUILD (hunk 5 above)
Comment (mattmoor) [reply BUILD]: removed

File: BUILD (hunk 2 above)
Comment (mattmoor) [reply BUILD:106]: I made these use the same helper function for both python versions with the shorter architecture list.

File: BUILD (hunk 3 above)
Comment (mattmoor) [reply BUILD:126]: added this to the loop, so we get the suffixes for these as well.

File: cloudbuild_docker.sh [hunk 6]
```diff
@@ -26,47 +26,25 @@ docker_manifest() {
...
-   gcr.io/$PROJECT_ID/python3:nonroot-amd64 \
-   gcr.io/$PROJECT_ID/python3:nonroot-arm64
-docker manifest push gcr.io/$PROJECT_ID/python3:nonroot
-
-docker manifest create gcr.io/$PROJECT_ID/python3:latest \
-   gcr.io/$PROJECT_ID/python3:latest-amd64 \
-   gcr.io/$PROJECT_ID/python3:latest-arm64
-docker manifest push gcr.io/$PROJECT_ID/python3:latest
+  docker_manifest gcr.io/$PROJECT_ID/cc${distro_suffix}:latest "amd64 arm arm64 s390x ppc64le"
+  docker_manifest gcr.io/$PROJECT_ID/cc${distro_suffix}:debug "amd64 arm arm64 s390x ppc64le"
```
Comment (chanseokoh) [inline cloudbuild_docker.sh:38]: Just realized the PR also publishes manifest lists for `nonroot` and `debug-nonroot` for `cc`.

```suggestion
  docker_manifest gcr.io/$PROJECT_ID/cc${distro_suffix}:nonroot "amd64 arm arm64 s390x ppc64le"
  docker_manifest gcr.io/$PROJECT_ID/cc${distro_suffix}:latest "amd64 arm arm64 s390x ppc64le"
  docker_manifest gcr.io/$PROJECT_ID/cc${distro_suffix}:debug-nonroot "amd64 arm arm64 s390x ppc64le"
  docker_manifest gcr.io/$PROJECT_ID/cc${distro_suffix}:debug "amd64 arm arm64 s390x ppc64le"
```

**Review by chanseokoh — COMMENTED** [review]
LGTM. One minor suggestion.

File: cloudbuild_docker.sh (hunk 6 above)
Comment (mattmoor) [reply cloudbuild_docker.sh]: added the nonroot variants for cc

**Review by chanseokoh — APPROVED** [review]
Thanks! Let's give it a spin.

**Pull request merged**